scripts/exact_1d_bayesian_filtering_demo.py

```diff
@@ -0,0 +1,515 @@
+        ax.set_ylabel("$x_k, y_k$")
+        ax.set_xlabel("k")
+        ax.legend(prop={"size": 16})
+
+    return x_grid, x_true, y
+
+
+# functions for the particle filter example
+
+# state transition function
+def state_trans_func1(x, v, k):
+    return x / 2 + 25 * x / (1 + x**2) + 8 * jnp.cos(1.2 * (k + 1)) + v
+
+
+# measurement function
+def measure_func1(x, e):
+    return x**2 / 20 + e
+
+
+# to get x from measurement without noise
+def inv_measure_func1(y):
+    x = jnp.sqrt(20 * y)
+    return [x, -x]
+
+
+# functions to get sample
+def v_rvs1(rng_key, shape):
+    return random.normal(rng_key, shape=shape) * jnp.sqrt(10)
+
+
+def e_rvs1(rng_key, shape):
+    return random.normal(rng_key, shape=shape)
+
+
+def x0_rvs1(rng_key, shape):
+    return random.normal(rng_key, shape=shape)
+
+
+# functions to get density
+v_pdf1 = functools.partial(jsp.stats.norm.pdf, scale=jnp.sqrt(10))
+e_pdf1 = functools.partial(jsp.stats.norm.pdf, scale=1)
+x0_pdf1 = jsp.stats.norm.pdf
+
+
+def the_particle_filter_example(
+    rng_key=random.PRNGKey(4),
+    grid_minval=-30,
+    grid_maxval=30,
+    num_grid_points=500,
+    max_iter=20,
+    iter_=14,
+    plot_all_densities=False):
+    # generate data points and densities
+    x_grid, x_true, y = experiment_setup(
+        rng_key=rng_key, grid_minval=grid_minval, grid_maxval=grid_maxval,
+        num_grid_points=num_grid_points, x0_rvs=x0_rvs1, v_rvs=v_rvs1,
+        e_rvs=e_rvs1, f=state_trans_func1, h=measure_func1,
+        max_iter=max_iter, plot_xy=False,
+    )
+
+
+    p_filter, p_pred, p_smooth = point_mass_density(
+        y, x_grid, x0_pdf1,
+        x_pdf=x_pdf, v_pdf=v_pdf1, e_pdf=e_pdf1,
+        f=state_trans_func1, h=measure_func1,
+    )
+
+    if plot_all_densities:
+        # looking for weird density plot by plotting all max_iter densities
+        plot_densities(
+            x_true, y, inv_measure_func1,
+            x_grid, p_pred, p_filter,
+            p_smooth, max_iter,
+        )
+
+    # plot the kth density
+    plot_density(
+        x_true, y, inv_measure_func1,
+        x_grid, p_pred, p_filter,
+        p_smooth, k=iter_, legend=True,
+        ax=None, title=f"Particle filter example densities at $x_{{{iter_}}}$",
+    )
+
+    plt.show()
+
+
+# functions for student t random walk example
+
+# state transition function
+def state_trans_func2(x, v, k=None):
+    return x + v
+
+
+# measurement function
+def measure_func2(x, e):
+    return x + e
+
+
+# to get x from measurement without noise
+def inv_measure_func2(y):
+    return y
+
+
+# functions to get sample
+def v_rvs2(rng_key, shape):
+    return random.t(rng_key, df=3, shape=shape)
+
+
+def e_rvs2(rng_key, shape):
+    return random.t(rng_key, df=3, shape=shape)
+
+
+def x0_rvs2(rng_key, shape):
+    return random.t(rng_key, df=3, shape=shape)
+
+
+# functions to get density
+pdf2 = functools.partial(jsp.stats.t.pdf, df=3)
+v_pdf2 = pdf2
+e_pdf2 = pdf2
+x0_pdf2 = pdf2
+
+
+def student_t_random_walk_example(
```

Comment:
Following lines 382:385, it's better if you add `v_pdf2`, `e_pdf2`, and `x0_pdf2` as parameters to your function. Then, when calling the function, you could do something like: `student_t_random_walk_example (..., v_pdf2=pdf2, e_pdf2=pdf2, x0_pdf2=pdf2)`

Reply:
Oh that sounds great. By the way I think that I need to use `student_t_random_walk_example (..., v_pdf=pdf2, e_pdf=pdf2, x0_pdf=pdf2)` instead, because `2` numbering is meaningless inside the example functions. On top of it, should I just declare internal functions `v_pdf`, `e_pdf`, `x0_pdf`, etc, in the example functions individually for encapsulation, or I should just use the existing practice, that define it outside and then pass it as argument?

e.g.
```
# state transition function
def state_trans_func2(x, v, k):
     return x / 2 + 25 * x / (1 + x**2) + 8 * jnp.cos(1.2 * (k + 1)) + v
...
def student_t_random_walk_example(
     ..., state_trans_func = state_trans_func2)

     ...
     state_trans_func()
```

vs

```
def student_t_random_walk_example(
     ...):

    def state_trans_func(x, v, k):
         return x / 2 + 25 * x / (1 + x**2) + 8 * jnp.cos(1.2 * (k + 1)) + v

    ...
    state_trans_func()
```